pmd-core/src/main/java/net/sourceforge/pmd/lang/metrics/AbstractMetricsFacade.java

```diff
@@ -0,0 +1,130 @@
+
+
+    /**
+     * Computes a metric identified by its code on a class AST node, possibly selecting a variant with the {@code
+     * MetricVersion} parameter.
+     *
+     * @param key     The key identifying the metric to be computed
+     * @param node    The node on which to compute the metric
+     * @param version The version of the metric
+     *
+     * @return The value of the metric, or {@code Double.NaN} if the value couln't be computed
+     */
+    public double computeForType(MetricKey<T> key, T node, MetricVersion version) {
+
+        checkKeyNotNull(key);
+
+        if (!key.supports(node)) {
+            return Double.NaN;
+        }
+
+        MetricVersion safeVersion = (version == null) ? Version.STANDARD : version;
+        MetricMemoizer<T> memoizer = getLanguageSpecificProjectMirror().getClassStats(node.getQualifiedName());
+
+        return memoizer == null ? Double.NaN
+                                : getLanguageSpecificComputer().computeForType(key, node, false, safeVersion, memoizer);
+    }
+
+
+    /**
+     * Computes a metric identified by its key on a operation AST node.
+     *
+     * @param key     The key identifying the metric to be computed
+     * @param node    The node on which to compute the metric
+     * @param version The version of the metric
+     *
+     * @return The value of the metric, or {@code Double.NaN} if the value couln't be computed
+     */
+    public double computeForOperation(MetricKey<O> key, O node,
+                                      MetricVersion version) {
+
+        checkKeyNotNull(key);
+
+        if (!key.supports(node)) {
+            return Double.NaN;
+        }
+
+        MetricVersion safeVersion = (version == null) ? Version.STANDARD : version;
+        MetricMemoizer<O> memoizer = getLanguageSpecificProjectMirror().getOperationStats(node.getQualifiedName());
+
+        return memoizer == null ? Double.NaN
+                                : getLanguageSpecificComputer().computeForOperation(key, node, false,
+                                                                                    safeVersion, memoizer);
+
+    }
+
+
+    /**
+     * Compute the sum, average, or highest value of the operation metric on all operations of the class node. The type
+     * of operation is specified by the {@link ResultOption} parameter.
+     *
+     * @param key     The key identifying the metric to be computed
+     * @param node    The node on which to compute the metric
+     * @param version The version of the metric
+     * @param option  The result option to use
+     *
+     * @return The value of the metric, or {@code Double.NaN} if the value couln't be computed or {@code option} is
+     * {@code null}
+     */
+    public double computeWithResultOption(MetricKey<O> key, T node,
+                                          MetricVersion version, ResultOption option) {
+
+        checkKeyNotNull(key);
```

Comment:
You could use for these null checks (key + option) the utility method java.util.Objects.requireNonNull, e.g.

    Objects.requireNonNull(key, "The metric key must not be null");
    Objects.requireNonNull(option, "The result option must not be null");